pallets/crowdloans/src/lib.rs

```diff
@@ -531,6 +554,78 @@ pub mod pallet {
     }
 
     impl<T: Config> Pallet<T> {
+        fn distribute_project_tokens(_block_number: T::BlockNumber) -> Result<(), Error<T>> {
+            let mut lock = StorageLock::<Time>::with_deadline(
+                b"distribute_project_tokens::lock",
+                Duration::from_millis(T::LockPeriod::get()),
+            );
+            if lock.try_lock().is_err() {
+                return Err(Error::<T>::GetLockFailed);
+            }
+            // AccountId, cTokenId, percentOfIssuance
+            let ownership_snapshot = Self::ctoken_ownership_snapshot();
+
+            // now for each account lets payout
+            for (account_id, accounts_ownership) in ownership_snapshot.iter() {
+                // payout for each vault
+                for (ctoken, percent_of_ctokens) in accounts_ownership.iter() {
+                    // get the total number of ptokens to be distributed
+                    let ptoken = Self::get_ptoken_from_ctoken(ctoken);
+                    let current_distribution_total = Self::get_distribution_amount(ptoken);
+
+                    // the user should recieve this portion of ptokens
+                    let amount = current_distribution_total * percent_of_ctokens;
+
+                    // transfer tokens to user from this account
+                    T::Assets::transfer(ptoken, &Self::account_id(), &account_id, amount, true)
+                        .map_err(|_: DispatchError| Error::<T>::InsufficientBalance)?;
+
+                    // emit event that we distributed
+                    Self::deposit_event(Event::<T>::DistributedProjectToken(
+                        ptoken,
+                        account_id.clone(),
+                        amount,
+                    ));
+                }
+            }
+            Ok(())
+        }
+
+        // TODO: how do we determine the ptoken for each ctoken
+        fn get_ptoken_from_ctoken(_ctoken: &u32) -> u32 {
+            0
+        }
+
+        // TODO: how do we determine the balance of tokens to distribute this iteration
+        fn get_distribution_amount(_ptoken: u32) -> u128 {
+            10_000
+        }
+
+        pub fn ctoken_ownership_snapshot() -> BTreeMap<T::AccountId, BTreeMap<u32, u128>> {
+            // iterate over all accounts on chain and build ownership map for each
+            frame_system::Account::<T>::iter()
```

Comment:
this is two much... I still prefer not using OCW because we had bad memory about OCW in liquidation pallet, it affected a lot the block production time